File: src/main/java/org/cubexmc/metro/lifecycle/CommandRegistration.java
package org.cubexmc.metro.lifecycle;

import java.lang.reflect.Proxy;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

import org.bukkit.Bukkit;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Entity;
import org.cubexmc.metro.model.EntityModelController;
import org.bukkit.entity.Player;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.command.newcmd.LineCommand;
import org.cubexmc.metro.command.newcmd.MetroMainCommand;
import org.cubexmc.metro.command.newcmd.PortalCommand;
import org.cubexmc.metro.command.newcmd.StopCommand;
import org.cubexmc.metro.manager.LineManager;
import org.cubexmc.metro.manager.PortalManager;
import org.cubexmc.metro.manager.StopManager;
import org.cubexmc.metro.service.StopCommandService;
import org.cubexmc.metro.util.VersionUtil;
import org.incendo.cloud.CommandManager;
import org.incendo.cloud.SenderMapper;
import org.incendo.cloud.annotations.AnnotationParser;
import org.incendo.cloud.bukkit.CloudBukkitCapabilities;
import org.incendo.cloud.context.CommandContext;
import org.incendo.cloud.context.CommandInput;
import org.incendo.cloud.execution.ExecutionCoordinator;
import org.incendo.cloud.paper.LegacyPaperCommandManager;
import org.incendo.cloud.paper.PaperCommandManager;
import org.incendo.cloud.suggestion.Suggestion;

/**
 * Registers Cloud commands and suggestion providers.
 */
public class CommandRegistration {

    private final Metro plugin;
    private final LineManager lineManager;
    private final StopManager stopManager;
    private final PortalManager portalManager;

    public CommandRegistration(Metro plugin, LineManager lineManager, StopManager stopManager, PortalManager portalManager) {
        this.plugin = plugin;
        this.lineManager = lineManager;
        this.stopManager = stopManager;
        this.portalManager = portalManager;
    }

    public Result register() {
        if (VersionUtil.isVersionAtLeast(26, 1, 0)) {
            try {
                new BukkitFallbackCommandRegistration(plugin, lineManager, stopManager, portalManager).register();
                plugin.getLogger().info("已为 Minecraft 26.1+ 启用 Bukkit 命令兼容层。");
                return new Result(null, null);
            } catch (Exception e) {
                plugin.getLogger().severe("Failed to initialize Bukkit command fallback:");
                e.printStackTrace();
                Bukkit.getPluginManager().disablePlugin(plugin);
                return null;
            }
        }

        try {
            CommandManager<CommandSender> commandManager = createCommandManager();
            AnnotationParser<CommandSender> annotationParser =
                    new AnnotationParser<>(commandManager, CommandSender.class);

            registerSuggestionProviders(commandManager);
            annotationParser.parse(
                    new MetroMainCommand(plugin, lineManager, stopManager),
                    new LineCommand(plugin, lineManager, stopManager),
                    new StopCommand(plugin, stopManager, lineManager),
                    new PortalCommand(plugin)
            );

            plugin.getLogger().info("Cloud Command Framework initialized successfully.");
            return new Result(commandManager, annotationParser);
        } catch (Exception | LinkageError e) {
            plugin.getLogger().warning("Cloud Command Framework 初始化失败，将尝试 Bukkit fallback。原因: "
                    + e.getClass().getSimpleName() + ": " + e.getMessage());
            try {
                new BukkitFallbackCommandRegistration(plugin, lineManager, stopManager, portalManager).register();
                return new Result(null, null);
            } catch (Exception fallbackError) {
                plugin.getLogger().severe("Failed to initialize Cloud Command Framework and Bukkit fallback:");
                fallbackError.printStackTrace();
                Bukkit.getPluginManager().disablePlugin(plugin);
                return null;
            }
        }
    }

    private CommandManager<CommandSender> createCommandManager() throws Exception {
        try {
            Class.forName("io.papermc.paper.plugin.lifecycle.event.LifecycleEventManager");
            final Class<?> commandSourceStackClass =
                    Class.forName("io.papermc.paper.command.brigadier.CommandSourceStack");

            @SuppressWarnings({"unchecked", "rawtypes"})
            SenderMapper<?, CommandSender> mapper = SenderMapper.create(
                    source -> {
                        try {
                            return (CommandSender) source.getClass().getMethod("getSender").invoke(source);
                        } catch (Exception e) {
                            throw new RuntimeException("Failed to map CommandSourceStack to CommandSender", e);
                        }
                    },
                    sender -> Proxy.newProxyInstance(
                            commandSourceStackClass.getClassLoader(),
                            new Class<?>[]{commandSourceStackClass},
                            (proxy, method, args) -> {
                                String name = method.getName();
                                if ("getSender".equals(name)) {
                                    return sender;
                                }
                                if ("getLocation".equals(name)) {
                                    return sender instanceof Entity entity ? entity.getLocation() : null;
                                }
                                if ("getExecutor".equals(name)) {
                                    return sender instanceof Entity ? sender : null;
                                }
                                if ("toString".equals(name)) {
                                    return "CommandSourceStackProxy[" + sender.getName() + "]";
                                }
                                if ("equals".equals(name)) {
                                    return args != null && args.length == 1 && proxy == args[0];
                                }
                                if ("hashCode".equals(name)) {
                                    return System.identityHashCode(proxy);
                                }
                                return null;
                            }
                    )
            );

            try {
                CommandManager<CommandSender> manager =
                        (CommandManager<CommandSender>) PaperCommandManager.builder((SenderMapper) mapper)
                                .executionCoordinator(ExecutionCoordinator.simpleCoordinator())
                                .buildOnEnable(plugin);
                plugin.getLogger().info("已加载新版 PaperCommandManager (1.20.5+)");
                return manager;
            } catch (RuntimeException | LinkageError e) {
                plugin.getLogger().warning("新版 PaperCommandManager 初始化失败，将降级为兼容命令注册。原因: "
                        + e.getClass().getSimpleName() + ": " + e.getMessage());
                return createLegacyCommandManager(false);
            }
        } catch (ClassNotFoundException e) {
            return createLegacyCommandManager(true);
        }
    }

    private LegacyPaperCommandManager<CommandSender> createLegacyCommandManager(boolean enableNativeBrigadier)
            throws Exception {
        LegacyPaperCommandManager<CommandSender> legacyManager =
                new LegacyPaperCommandManager<>(
                        plugin,
                        ExecutionCoordinator.simpleCoordinator(),
                        SenderMapper.identity()
                );

        if (enableNativeBrigadier && legacyManager.hasCapability(CloudBukkitCapabilities.NATIVE_BRIGADIER)) {
            try {
                legacyManager.registerBrigadier();
            } catch (RuntimeException | LinkageError e) {
                plugin.getLogger().warning("Legacy Brigadier 注册失败，将继续使用 Bukkit 命令。原因: "
                        + e.getClass().getSimpleName() + ": " + e.getMessage());
            }
        }
        if (legacyManager.hasCapability(CloudBukkitCapabilities.ASYNCHRONOUS_COMPLETION)) {
            try {
                legacyManager.registerAsynchronousCompletions();
            } catch (RuntimeException | LinkageError e) {
                plugin.getLogger().warning("异步命令补全注册失败，将继续使用同步补全。原因: "
                        + e.getClass().getSimpleName() + ": " + e.getMessage());
            }
        }

        plugin.getLogger().info(enableNativeBrigadier
                ? "已加载兼容版 LegacyPaperCommandManager (1.20.4 及以下)"
                : "已加载降级版 LegacyPaperCommandManager (Bukkit command fallback)");
        return legacyManager;
    }

    private void registerSuggestionProviders(CommandManager<CommandSender> commandManager) {
        commandManager.parserRegistry().registerSuggestionProvider("lineIds",
                (context, input) -> toSuggestionsFuture(lineIdSuggestions(context, input)));
        commandManager.parserRegistry().registerSuggestionProvider("stopIds",
                (context, input) -> toSuggestionsFuture(stopIdSuggestions(context, input)));
        commandManager.parserRegistry().registerSuggestionProvider("portalIds",
                (context, input) -> toSuggestionsFuture(portalIdSuggestions(context, input)));
        commandManager.parserRegistry().registerSuggestionProvider("playerNames",
                (context, input) -> toSuggestionsFuture(playerNameSuggestions(context, input)));
        commandManager.parserRegistry().registerSuggestionProvider("players",
                (context, input) -> toSuggestionsFuture(playerNameSuggestions(context, input)));
        commandManager.parserRegistry().registerSuggestionProvider("lineColors",
                (context, input) -> toSuggestionsFuture(List.of(
                        "&0", "&1", "&2", "&3", "&4", "&5", "&6", "&7",
                        "&8", "&9", "&a", "&b", "&c", "&d", "&e", "&f",
                        "&#55AAFF")));
        commandManager.parserRegistry().registerSuggestionProvider("protectModes",
                (context, input) -> toSuggestionsFuture(List.of(
                        "status", "on", "off", "enable", "disable", "enabled", "disabled", "true", "false")));
        commandManager.parserRegistry().registerSuggestionProvider("titleTypes",
                (context, input) -> toSuggestionsFuture(StopCommandService.TITLE_TYPES.stream().sorted().toList()));
        commandManager.parserRegistry().registerSuggestionProvider("titleKeys",
                (context, input) -> toSuggestionsFuture(StopCommandService.TITLE_KEYS.stream().sorted().toList()));
        commandManager.parserRegistry().registerSuggestionProvider("linkActions",
                (context, input) -> toSuggestionsFuture(List.of("allow", "deny")));
        commandManager.parserRegistry().registerSuggestionProvider("pageNumbers",
                (context, input) -> toSuggestionsFuture(List.of("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")));
        commandManager.parserRegistry().registerSuggestionProvider("stopIndexes",
                (context, input) -> toSuggestionsFuture(List.of("0", "1", "2", "3", "4", "5", "10")));
        commandManager.parserRegistry().registerSuggestionProvider("yawValues",
                (context, input) -> toSuggestionsFuture(List.of("0", "90", "180", "-90")));
        commandManager.parserRegistry().registerSuggestionProvider("speedValues",
                (context, input) -> toSuggestionsFuture(List.of("0.4", "0.8", "1.0", "1.2")));
        commandManager.parserRegistry().registerSuggestionProvider("priceValues",
                (context, input) -> toSuggestionsFuture(List.of("0", "1", "2", "5", "10")));
        commandManager.parserRegistry().registerSuggestionProvider("priceModes",
                (context, input) -> toSuggestionsFuture(List.of("flat", "distance", "interval")));
        commandManager.parserRegistry().registerSuggestionProvider("lineStatusValues",
                (context, input) -> toSuggestionsFuture(List.of("normal", "suspended", "maintenance")));
        commandManager.parserRegistry().registerSuggestionProvider("trainControlModes",
                (context, input) -> toSuggestionsFuture(List.of("reactive", "kinematic", "leashed", "default")));
        commandManager.parserRegistry().registerSuggestionProvider("entityTypes",
                (context, input) -> toSuggestionsFuture(EntityModelController.suggestedEntityTypeNames()));
    }

    private Iterable<String> lineIdSuggestions(CommandContext<CommandSender> context, CommandInput input) {
        return lineManager.getAllLines().stream()
                .map(org.cubexmc.metro.model.Line::getId)
                .toList();
    }

    private Iterable<String> stopIdSuggestions(CommandContext<CommandSender> context, CommandInput input) {
        return new ArrayList<>(stopManager.getAllStopIds());
    }

    private Iterable<String> portalIdSuggestions(CommandContext<CommandSender> context, CommandInput input) {
        return portalManager.getAllPortals().stream()
                .map(org.cubexmc.metro.model.Portal::getId)
                .sorted()
                .toList();
    }

    private Iterable<String> playerNameSuggestions(CommandContext<CommandSender> context, CommandInput input) {
        return Bukkit.getOnlinePlayers().stream()
                .map(Player::getName)
                .sorted()
                .toList();
    }

    private CompletableFuture<? extends Iterable<? extends Suggestion>> toSuggestionsFuture(Iterable<String> values) {
        ArrayList<Suggestion> suggestions = new ArrayList<>();
        for (String value : values) {
            suggestions.add(Suggestion.suggestion(value));
        }
        return CompletableFuture.completedFuture(suggestions);
    }

    public record Result(CommandManager<CommandSender> commandManager,
                         AnnotationParser<CommandSender> annotationParser) {
    }
}

File: src/main/java/org/cubexmc/metro/lifecycle/CommandRegistration.kt
package org.cubexmc.metro.lifecycle

import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import org.bukkit.entity.Entity
import org.cubexmc.metro.Metro
import org.cubexmc.metro.command.newcmd.LineCommand
import org.cubexmc.metro.command.newcmd.MetroMainCommand
import org.cubexmc.metro.command.newcmd.PortalCommand
import org.cubexmc.metro.command.newcmd.StopCommand
import org.cubexmc.metro.manager.LineManager
import org.cubexmc.metro.manager.PortalManager
import org.cubexmc.metro.manager.StopManager
import org.cubexmc.metro.model.EntityModelController
import org.cubexmc.metro.service.StopCommandService
import org.cubexmc.metro.util.VersionUtil
import org.incendo.cloud.CommandManager
import org.incendo.cloud.SenderMapper
import org.incendo.cloud.annotations.AnnotationParser
import org.incendo.cloud.bukkit.CloudBukkitCapabilities
import org.incendo.cloud.execution.ExecutionCoordinator
import org.incendo.cloud.paper.LegacyPaperCommandManager
import org.incendo.cloud.suggestion.Suggestion
import java.lang.reflect.Proxy
import java.util.concurrent.CompletableFuture

/**
 * Registers Cloud commands and suggestion providers.
 */
class CommandRegistration(
    private val plugin: Metro,
    private val lineManager: LineManager,
    private val stopManager: StopManager,
    private val portalManager: PortalManager,
) {

    fun register(): Result? {
        if (VersionUtil.isVersionAtLeast(FALLBACK_MAJOR, FALLBACK_MINOR, FALLBACK_PATCH)) {
            try {
                BukkitFallbackCommandRegistration(plugin, lineManager, stopManager, portalManager).register()
                plugin.logger.info("已为 Minecraft 26.1+ 启用 Bukkit 命令兼容层。")
                return Result(null, null)
            } catch (e: Exception) {
                plugin.logger.severe("Failed to initialize Bukkit command fallback:")
                e.printStackTrace()
                Bukkit.getPluginManager().disablePlugin(plugin)
                return null
            }
        }

        return try {
            val commandManager = createCommandManager()
            val annotationParser = AnnotationParser(commandManager, CommandSender::class.java)

            registerSuggestionProviders(commandManager)
            annotationParser.parse(
                MetroMainCommand(plugin, lineManager, stopManager),
                LineCommand(plugin, lineManager, stopManager),
                StopCommand(plugin, stopManager, lineManager),
                PortalCommand(plugin),
            )

            plugin.logger.info("Cloud Command Framework initialized successfully.")
            Result(commandManager, annotationParser)
        } catch (e: Exception) {
            registerBukkitFallbackAfterCloudFailure(e)
        } catch (e: LinkageError) {
            registerBukkitFallbackAfterCloudFailure(e)
        }
    }

    private fun registerBukkitFallbackAfterCloudFailure(cause: Throwable): Result? {
        plugin.logger.warning(
            "Cloud Command Framework 初始化失败，将尝试 Bukkit fallback。原因: " +
                cause.javaClass.simpleName + ": " + cause.message,
        )
        return try {
            BukkitFallbackCommandRegistration(plugin, lineManager, stopManager, portalManager).register()
            Result(null, null)
        } catch (fallbackError: Exception) {
            plugin.logger.severe("Failed to initialize Cloud Command Framework and Bukkit fallback:")
            fallbackError.printStackTrace()
            Bukkit.getPluginManager().disablePlugin(plugin)
            null
        }
    }

    private fun createCommandManager(): CommandManager<CommandSender> {
        try {
            Class.forName("io.papermc.paper.plugin.lifecycle.event.LifecycleEventManager")
            val commandSourceStackClass = Class.forName("io.papermc.paper.command.brigadier.CommandSourceStack")

            val mapper: SenderMapper<Any, CommandSender> =
                SenderMapper.create(
                    { source ->
                        try {
                            source.javaClass.getMethod("getSender").invoke(source) as CommandSender
                        } catch (e: Exception) {
                            throw RuntimeException("Failed to map CommandSourceStack to CommandSender", e)
                        }
                    },
                    { sender ->
                        Proxy.newProxyInstance(
                            commandSourceStackClass.classLoader,
                            arrayOf<Class<*>>(commandSourceStackClass),
                        ) { proxy, method, args ->
                            when (method.name) {
                                "getSender" -> sender
                                "getLocation" -> (sender as? Entity)?.location
                                "getExecutor" -> if (sender is Entity) sender else null
                                "toString" -> "CommandSourceStackProxy[" + sender.name + "]"
                                "equals" -> args != null && args.size == 1 && proxy === args[0]
                                "hashCode" -> System.identityHashCode(proxy)
                                else -> null
                            }
                        }
                    },
                )

            return try {
                val manager = PaperCommandManagerBootstrap.buildOnEnable(plugin, mapper)
                plugin.logger.info("已加载新版 PaperCommandManager (1.20.5+)")
                manager
            } catch (e: RuntimeException) {
                warnModernManagerFailure(e)
                createLegacyCommandManager(false)
            } catch (e: LinkageError) {
                warnModernManagerFailure(e)
                createLegacyCommandManager(false)
            }
        } catch (_: ClassNotFoundException) {
            return createLegacyCommandManager(true)
        }
    }

    private fun warnModernManagerFailure(cause: Throwable) {
        plugin.logger.warning(
            "新版 PaperCommandManager 初始化失败，将降级为兼容命令注册。原因: " +
                cause.javaClass.simpleName + ": " + cause.message,
        )
    }

    private fun createLegacyCommandManager(enableNativeBrigadier: Boolean): LegacyPaperCommandManager<CommandSender> {
        val legacyManager =
            LegacyPaperCommandManager(
                plugin,
                ExecutionCoordinator.simpleCoordinator(),
                SenderMapper.identity<CommandSender>(),
            )

        if (enableNativeBrigadier && legacyManager.hasCapability(CloudBukkitCapabilities.NATIVE_BRIGADIER)) {
            try {
                legacyManager.registerBrigadier()
            } catch (e: RuntimeException) {
                warnBrigadierFailure(e)
            } catch (e: LinkageError) {
                warnBrigadierFailure(e)
            }
        }
        if (legacyManager.hasCapability(CloudBukkitCapabilities.ASYNCHRONOUS_COMPLETION)) {
            try {
                legacyManager.registerAsynchronousCompletions()
            } catch (e: RuntimeException) {
                warnAsynchronousCompletionFailure(e)
            } catch (e: LinkageError) {
                warnAsynchronousCompletionFailure(e)
            }
        }

        plugin.logger.info(
            if (enableNativeBrigadier) {
                "已加载兼容版 LegacyPaperCommandManager (1.20.4 及以下)"
            } else {
                "已加载降级版 LegacyPaperCommandManager (Bukkit command fallback)"
            },
        )
        return legacyManager
    }

    private fun warnBrigadierFailure(cause: Throwable) {
        plugin.logger.warning(
            "Legacy Brigadier 注册失败，将继续使用 Bukkit 命令。原因: " +
                cause.javaClass.simpleName + ": " + cause.message,
        )
    }

    private fun warnAsynchronousCompletionFailure(cause: Throwable) {
        plugin.logger.warning(
            "异步命令补全注册失败，将继续使用同步补全。原因: " +
                cause.javaClass.simpleName + ": " + cause.message,
        )
    }

    private fun registerSuggestionProviders(commandManager: CommandManager<CommandSender>) {
        val registry = commandManager.parserRegistry()
        registry.registerSuggestionProvider("lineIds") { _, _ -> toSuggestionsFuture(lineIdSuggestions()) }
        registry.registerSuggestionProvider("stopIds") { _, _ -> toSuggestionsFuture(stopIdSuggestions()) }
        registry.registerSuggestionProvider("portalIds") { _, _ -> toSuggestionsFuture(portalIdSuggestions()) }
        registry.registerSuggestionProvider("playerNames") { _, _ -> toSuggestionsFuture(playerNameSuggestions()) }
        registry.registerSuggestionProvider("players") { _, _ -> toSuggestionsFuture(playerNameSuggestions()) }
        registry.registerSuggestionProvider("lineColors") { _, _ -> toSuggestionsFuture(LINE_COLORS) }
        registry.registerSuggestionProvider("protectModes") { _, _ -> toSuggestionsFuture(PROTECT_MODES) }
        registry.registerSuggestionProvider("titleTypes") { _, _ ->
            toSuggestionsFuture(StopCommandService.TITLE_TYPES.sorted())
        }
        registry.registerSuggestionProvider("titleKeys") { _, _ ->
            toSuggestionsFuture(StopCommandService.TITLE_KEYS.sorted())
        }
        registry.registerSuggestionProvider("linkActions") { _, _ -> toSuggestionsFuture(listOf("allow", "deny")) }
        registry.registerSuggestionProvider("pageNumbers") { _, _ -> toSuggestionsFuture(PAGE_NUMBERS) }
        registry.registerSuggestionProvider("stopIndexes") { _, _ -> toSuggestionsFuture(STOP_INDEXES) }
        registry.registerSuggestionProvider("yawValues") { _, _ -> toSuggestionsFuture(YAW_VALUES) }
        registry.registerSuggestionProvider("speedValues") { _, _ -> toSuggestionsFuture(SPEED_VALUES) }
        registry.registerSuggestionProvider("priceValues") { _, _ -> toSuggestionsFuture(PRICE_VALUES) }
        registry.registerSuggestionProvider("priceModes") { _, _ -> toSuggestionsFuture(PRICE_MODES) }
        registry.registerSuggestionProvider("lineStatusValues") { _, _ -> toSuggestionsFuture(LINE_STATUS_VALUES) }
        registry.registerSuggestionProvider("trainControlModes") { _, _ ->
            toSuggestionsFuture(listOf("reactive", "kinematic", "leashed", "default"))
        }
        registry.registerSuggestionProvider("entityTypes") { _, _ ->
            toSuggestionsFuture(EntityModelController.suggestedEntityTypeNames())
        }
    }

    private fun lineIdSuggestions(): List<String> = lineManager.getAllLines().map { it.id }

    private fun stopIdSuggestions(): List<String> = ArrayList(stopManager.getAllStopIds())

    private fun portalIdSuggestions(): List<String> = portalManager.getAllPortals().map { it.id }.sorted()

    private fun playerNameSuggestions(): List<String> = Bukkit.getOnlinePlayers().map { it.name }.sorted()

    private fun toSuggestionsFuture(values: Iterable<String>): CompletableFuture<List<Suggestion>> =
        CompletableFuture.completedFuture(values.map { Suggestion.suggestion(it) })

    data class Result(
        private val commandManager: CommandManager<CommandSender>?,
        private val annotationParser: AnnotationParser<CommandSender>?,
    ) {
        fun commandManager(): CommandManager<CommandSender>? = commandManager

        fun annotationParser(): AnnotationParser<CommandSender>? = annotationParser
    }

    private companion object {
        const val FALLBACK_MAJOR = 26
        const val FALLBACK_MINOR = 1
        const val FALLBACK_PATCH = 0

        val LINE_COLORS =
            listOf(
                "&0", "&1", "&2", "&3", "&4", "&5", "&6", "&7",
                "&8", "&9", "&a", "&b", "&c", "&d", "&e", "&f",
                "&#55AAFF",
            )
        val PROTECT_MODES =
            listOf("status", "on", "off", "enable", "disable", "enabled", "disabled", "true", "false")
        val PAGE_NUMBERS = listOf("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")
        val STOP_INDEXES = listOf("0", "1", "2", "3", "4", "5", "10")
        val YAW_VALUES = listOf("0", "90", "180", "-90")
        val SPEED_VALUES = listOf("0.4", "0.8", "1.0", "1.2")
        val PRICE_VALUES = listOf("0", "1", "2", "5", "10")
        val PRICE_MODES = listOf("flat", "distance", "interval")
        val LINE_STATUS_VALUES = listOf("normal", "suspended", "maintenance")
    }
}
